File: algorithms/2_medium/E/00809_expressive_words/solution.c
inline static bool is_sub(const char *s, const size_t s_len, const char *t) {
    if (t == NULL) { return false; }

    const size_t t_len = strlen(t);
    size_t s_idx = 0;
    size_t t_idx = 0;

    while (s_idx < s_len && t_idx < t_len) {
        if (s_len - s_idx < t_len - t_idx || s[s_idx] != t[t_idx]) { return false; }

        const char ch = s[s_idx];
        const size_t s_begin = s_idx;
        const size_t t_begin = t_idx;

        while (s_idx < s_len && s[s_idx] == ch) {
            s_idx += 1;
        }

        while (t_idx < t_len && t[t_idx] == ch) {
            t_idx += 1;
        }

        const size_t s_cnt = s_idx - s_begin;
        const size_t t_cnt = t_idx - t_begin;

        if (s_cnt != t_cnt && (s_cnt < t_cnt || s_cnt == 2)) {
            return false;
        } 
    }

    return s_idx == s_len && t_idx == t_len;
}

int expressiveWords(const char *s, char** words, const int words_len) {
    if (s == NULL || s[0] == '\0' || words == NULL || words_len <= 0) { return -1; }

    int cnt = 0;
    const size_t s_len = strlen(s);

    for (int i = 0; i < words_len; i += 1) {
        if (is_sub(s, s_len, words[i])) {
            cnt += 1;
        }
    }

    return cnt;
}

File: algorithms/2_medium/E/00809_expressive_words/solution.rs
impl Solution {
    #[inline(always)]
    fn is_sub(s: &[u8], t: &[u8]) -> bool {
        let (s_len, t_len) = (s.len(), t.len());
        let (mut s_idx, mut t_idx) = (0_usize, 0_usize);

        while s_idx < s_len && t_idx < t_len {
            if s_len - s_idx < t_len - t_idx || s[s_idx] != t[t_idx] { return false; }

            let byte = s[s_idx];
            let (s_begin, t_begin) = (s_idx, t_idx);

            while s_idx < s_len && s[s_idx] == byte {
                s_idx += 1;
            }

            while t_idx < t_len && t[t_idx] == byte {
                t_idx += 1;
            }

            let s_cnt = s_idx - s_begin;
            let t_cnt = t_idx - t_begin;

            if s_cnt != t_cnt && (s_cnt < t_cnt || s_cnt == 2) {
                return false;
            } 
        }

        s_idx == s_len && t_idx == t_len
    }

    pub fn expressive_words(s: String, words: Vec<String>) -> i32 {
        let s_bytes = s.as_bytes();
        let mut cnt = 0_i32;

        for w in words {
            if Self::is_sub(s_bytes, w.as_bytes()) {
                cnt += 1;
            }
        }

        cnt
    }
}
